### DISCUSSION TITLE
How to change Fast-DDS XML configuration to communicate over network via wlp58s0 (different network interface)

### DISCUSSION THREAD

**Discussion by ljaniec** [eProsima/Fast-DDS#3550] [Q&A]
Hello,

I am working on upgrading TurtleBot2 from ROS 2 Galactic with Cyclone DDS to ROS 2 Humble with Fast-DDS as middleware. My problem is configuring Fast-DDS to communicate over the network via wlp58s0, rather than trying to communicate over an RJ-45 cable (this is what the lidar on the robot is connected to). In ROS 2 Galactic there was another middleware, Cyclone DDS, and there it was easier, with an example straight from the documentation (below).

    <?xml version="1.0" encoding="UTF-8" ?>
    <CycloneDDS xmlns="https://cdds.io/config" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="https://cdds.io/config https://raw.githubusercontent.com/eclipse-cyclonedds/cyclonedds/master/etc/cyclonedds.xsd">
        <Domain id="any">
            <General>
                <NetworkInterfaceAddress>wlp58s0</NetworkInterfaceAddress>
            </General>
        </Domain>
    </CycloneDDS>

 In ROS 2 Humble the default is Fast-DDS, which did not have such clear documentation for us. Unfortunately, I don't really know what would be the equivalent of an analogous change. When I read the Fast-DDS documentation, I don't really know how to do it in this case; I found some links regarding ROS 2 in the documentation:
 
- https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/ros2_configure.html

- https://fast-dds.docs.eprosima.com/en/latest/fastdds/xml_configuration/xml_configuration.html

Should I do this using whitelists?

- https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/whitelist.html

Do I need to create a profile (separate for TCP and UDP, I think) with a whitelist of IP addresses assigned to those interfaces over which Fast DDS can communicate? Presumably 127.0.0.1 and the IP address of the Wi-Fi interface? What would the XML of such a configuration look like?

Best regards,

Łukasz Janiec

PS I posted this as a question on [ROS Answers](https://answers.ros.org/question/415693/how-to-change-fast-dds-xml-configuration-to-communicate-over-network-via-wlp58s0-different-network-interface/), without any hint, so I am trying to find something here.

**Comment by JesusPoderoso** [comment]
Hello @ljaniec 

>Should I do this using whitelists?

You are right, the equivalent would be including the IP address(es) of the node(s)  you want to communicate with.

The XML example displayed in the [Interface Whitelist Fast DDS documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/whitelist.html) should suit you.

(Make sure to include `is_default_profile="true"` in the participant attributes to set it automatically).

```
<?xml version="1.0" encoding="UTF-8" ?>
<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
    <transport_descriptors>
        <transport_descriptor>
            <transport_id>CustomUdpTransport</transport_id>
            <type>UDPv4</type>
            <interfaceWhiteList>
                <address>127.0.0.1</address>
            </interfaceWhiteList>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="CustomUdpTransportParticipant" is_default_profile="true">
        <rtps>
            <useBuiltinTransports>false</useBuiltinTransports>
            <userTransports>
                <transport_id>CustomUdpTransport</transport_id>
            </userTransports>
        </rtps>
    </participant>
</profiles>
```

**Reply by ljaniec** [reply]
Ok, I will try this today in the lab and report later.
The setup I am planning to do looks like this:

- one main PC (Ubuntu 22.04) with connection to the motion capture Motive 3 through MOCAP4ROS2 package (Windows, separate PC), it will have the top layer of control of multiple mobile robots, with static IP of `192.168.5.50`
- 6x Turtlebots 2 with Intel NUC (Ubuntu 22.04), each with separate static IP of the form `192.168.5.5x`, `x` is the robot number, 1..6 

Then my XML should look like this, right?

```
<?xml version="1.0" encoding="UTF-8" ?>
<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
    <transport_descriptors>
        <transport_descriptor>
            <transport_id>CustomUdpTransport</transport_id>
            <type>UDPv4</type>
            <interfaceWhiteList>
                <address>192.168.5.51</address>
                <address>192.168.5.52</address>
                <address>192.168.5.53</address>
                <address>192.168.5.54</address>
                <address>192.168.5.55</address>
                <address>192.168.5.56</address>
            </interfaceWhiteList>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="CustomUdpTransportParticipant" is_default_profile="true">
        <rtps>
            <useBuiltinTransports>false</useBuiltinTransports>
            <userTransports>
                <transport_id>CustomUdpTransport</transport_id>
            </userTransports>
        </rtps>
    </participant>
</profiles>
```
Please correct me if I am wrong.

**Reply by ljaniec** [reply]
@JesusPoderoso 
I tried it with a TB2 bringup and Nav2 launch files and I got this type of errors:
```
[rviz2-1] 2023-05-31 19:44:29.807 [XMLPARSER Error] Error opening '/home/user/DEFAULT_FASTRTPS_PROFILES.xml' -> Function loadXML
[rviz2-1] 2023-05-31 19:44:29.807 [XMLPARSER Error] Error parsing '/home/user/DEFAULT_FASTRTPS_PROFILES.xml' -> Function loadXMLFile
```
It showed sometimes with normal `ros2 topic list`, but `ros2 daemon stop` & `ros2 daemon start` helped. How can I fix errors encountered during launches? XML is (in theory) valid, I checked it with online validator, but maybe my setup requires different configuration? I heard that Fast Discovery Server is also a good option for this type of setup, what do you think?

**Comment by JesusPoderoso** [comment]
Hi @ljaniec 
I was not able to reproduce the XML parser error. In the mean time, could you describe the purpose of the deployment?
I suspect that maybe the interface whitelist may not be the configuration your deployment requires.
Is that XML configuration file being use only in the main PC, or is it configuring also the Turtlebots? Do you want the turtlebots to communicate with each other?

**Reply by ljaniec** [reply]
Hi @JesusPoderoso 
This is the picture in Drawio of the setup I wanted to create:
![schemat_lab07 drawio](https://github.com/eProsima/Fast-DDS/assets/11331846/ab3eaed8-3b52-4a3d-ab75-7b889d3a6eeb)
Each of the robots should be able to see and communicate over other robots topics etc. in their namespaces and the main PC with control system + it's topics.
I had the problem of proper configuration of the Turtlebot2 with FastDDS, because if I SSH'ed to the robot with turned on LiDAR, their topics stop showing on other robots (e.g. with simply `ros2 topic list` - they are there if we don't use `ssh` connection and lidar is on). The solution of this problem on the ROS 2 Galactic (with Cyclone DDS) was a proper XML with `NetworkInterfaceAddress` set to `wlp58s0`. We need to connect over SSH to the robots in order to enable the software there (Kobuki base, ROS2 nodes, especially Nav2 etc.).

I want to have a FastDDS XML configuration, that can be different on each of the robots and the PC, but I want to know how to work with it to allow/disallow connection between robots (e.g. in experiments with two robot groups working together without direct communication) in the future.

I am open to any suggestions how to solve this challenge, nothing is set in stone yet, so I can work on a better solution.

EDIT: Regarding the XML parser error - perhaps it was caused by the use of ROS2 components in the Nav2, described here: https://github.com/ros-planning/navigation2/tree/humble/nav2_bringup

**Comment by JesusPoderoso** [comment]
Hi @ljaniec 

In my opinion, the most dynamic and scalable solution for your deployment would be using a [discovery server](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/wifi/discovery_server_use_case.html) instead of whitelist interfaces.
You can run the server on your PC, and configure all the turtlebot2 with the server information. 

Here it is a [complete tutorial and demo for ROS2](https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/discovery_server/ros2_discovery_server.html).

If your are interested on a deep study, please do not hesitate to contact the [eProsima commercial support](https://github.com/eProsima/Fast-DDS#getting-help).
